### PR TITLE
Storage key failover logic

### PR DESCRIPTION
**Pull request by jwallra**
Open for feedback.  Still need to write/update unit tests.

### CODE REVIEW

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/crypto/CryptoUtils.java [hunk 1]
```diff
@@ -497,4 +512,12 @@ public String getNewEncryptedData() {
             return mNewEncryptedData;
         }
     }
+
+    /**
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/crypto/CryptoUtils.java]: remove public and javadoc

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 2]
```diff
@@ -729,6 +732,23 @@ else if (mUnknownSourcesDialog != null) {
...
+        /* Decrypt token if any. */
+        if (updateToken != null) {
+            CryptoUtils.DecryptedData decryptedData = CryptoUtils.getInstance(mContext).decrypt(updateToken, mobileCenterFailover);
+            String newEncryptedData = decryptedData.getNewEncryptedData();
+
+             /* Store new encrypted value if updated. */
+            if (newEncryptedData != null) {
+                PreferencesStorage.putString(PREFERENCE_KEY_UPDATE_TOKEN, newEncryptedData);
+            }
+            updateToken = decryptedData.getDecryptedData();
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]:
```
if (mobileCenterFailover)  {
    
     /* Store the token from Mobile Center into the new App Center storage, re-encrypting it */
     ...

     /* Also migrate store distribution group from old to new storage */
     ...
}
```

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 3]
```diff
@@ -868,11 +900,11 @@ public void onCallSucceeded(final String payload) {
...
-                        try {
-                            handleApiCallSuccess(releaseCallId, payload, ReleaseDetails.parse(payload));
-                        } catch (JSONException e) {
-                            onCallFailed(e);
-                        }
+                    try {
+                        handleApiCallSuccess(releaseCallId, payload, ReleaseDetails.parse(payload));
+                    } catch (JSONException e) {
+                        onCallFailed(e);
+                    }
```
Comment (jwallra) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: This was appearing as double-tabbed in my editor...?

Comment (guperrot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Not sure what you mean. If it's a formatting issue, feel free to reformat using Android Studio and commit.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 4]
```diff
@@ -714,13 +713,13 @@ else if (mUnknownSourcesDialog != null) {
...
             String distributionGroupId = PreferencesStorage.getString(PREFERENCE_KEY_DISTRIBUTION_GROUP_ID);
             if (updateToken != null || distributionGroupId != null) {
-
                 decryptAndGetReleaseDetails(updateToken, distributionGroupId, false);
                 return;
             }
             else {
 
                 /* Use failover logic to search for missing token/distribution group */
+                System.out.println("mmcps: " + mMobileCenterPreferenceStorage);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Please remove

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java [hunk 5]
```diff
@@ -135,6 +143,8 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
         /* Default download id when not found. */
         when(PreferencesStorage.getLong(PREFERENCE_KEY_DOWNLOAD_ID, INVALID_DOWNLOAD_IDENTIFIER)).thenReturn(INVALID_DOWNLOAD_IDENTIFIER);
 
+        mMobileCenterPreferencesStorage = mock(SharedPreferences.class);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java]: redundant with @Mock

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 6]
```diff
@@ -144,6 +149,13 @@ private void showUpdateSetupFailedDialog(ArgumentCaptor<DialogInterface.OnClickL
         PreferencesStorage.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY);
     }
 
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        when(mContext.getSharedPreferences(PREFERENCES_NAME_MOBILE_CENTER, Context.MODE_PRIVATE)).thenReturn(mMobileCenterPreferencesStorage);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: Looks like you defined mMobileCenterPreferencesStorage in Abstract test already, you can remove this method, the field in this class and probably move the the when to parent set up.

Either that or review why mMobileCenterPreferencesStorage is also in Abstract test.

**Pull request merged**